### PR TITLE
simplify get_class_in_module and fix for paths containing a dot

### LINKED ISSUE #29251: AutoModel.from_pretrained('.') raises ModuleNotFoundError with trust_remote_code=True
### System Info

- `transformers` version: 4.39.0.dev0 (commit c8d98405a)
- Platform: Linux-6.6.10-zen1-1-zen-cebtenzzre-x86_64-with-glibc2.39
- Python version: 3.11.3
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.2.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

This issue was reported in https://github.com/huggingface/transformers/issues/28919#issuecomment-1939195589 but not fixed by #29175.

1. `git clone https://huggingface.co/nomic-ai/nomic-embed-text-v1`
2. `cd nomic-embed-text-v1`
3. `git lfs pull` (if needed)
4. Try to load the locally downloaded model:
```
$ python -c 'from transformers import AutoModel; AutoModel.from_pretrained(".", trust_remote_code=True)'
Traceback (most recent call last):
  File "/home/jared/src/clones/transformers/src/transformers/dynamic_module_utils.py", line 203, in get_class_in_module
    module = importlib.import_module(module_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1142, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'transformers_modules.'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jared/src/clones/transformers/src/transformers/models/auto/auto_factory.py", line 521, in from_pretrained
    config, kwargs = AutoConfig.from_pretrained(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jared/src/clones/transformers/src/transformers/models/auto/configuration_auto.py", line 1120, in from_pretrained
    config_class = get_class_from_dynamic_module(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jared/src/clones/transformers/src/transformers/dynamic_module_utils.py", line 516, in get_class_from_dynamic_module
    return get_class_in_module(repo_id, class_name, final_module.replace(".py", ""))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jared/src/clones/transformers/src/transformers/dynamic_module_utils.py", line 215, in get_class_in_module
    module = importlib.machinery.SourceFileLoader(module_path, corrected_path).load_module()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 605, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 1120, in load_module
  File "<frozen importlib._bootstrap_external>", line 945, in load_module
  File "<frozen importlib._bootstrap>", line 290, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 721, in _load
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 936, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1073, in get_code
  File "<frozen importlib._bootstrap_external>", line 1130, in get_data
FileNotFoundError: [Errno 2] No such file or directory: '.home.jared..cache.huggingface.modules.transformers_modules...configuration_hf_nomic_bert.py'
```

### Expected behavior

I would expect `"."` to be interpreted as the path to the model, and for the model to successfully load.

Or at least a clear error message explaining what I am doing incorrectly - the unclear nature of the traceback makes this look like a bug to me.

### PR DESCRIPTION
**Pull request by cebtenzzre**
# What does this PR do?

A simpler implementation of get_class_in_module that also works when the path starts with e.g. `./` or otherwise contains dot characters *outside* of the repo ID.

Fixes #29251

I should write a test. The test in #29175 didn't fail before the PR (with code from commit 89c64817c), so it does not seem valid.

This does change the API of get_class_in_module to expect the path to an actual file, with the extension - but since the previous PR changed the API already, this seems fine.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/dynamic_module_utils.py
Comment (not-lain) [inline src/transformers/dynamic_module_utils.py:188]: could you return the original parameter `repo_id` also the documentation for it too, changing parameter names or removing them should not be be done this fast, usually the maintainers deperacate a parameter after a big release of the library in case other libraries or modules use this function too, so it's too much effort.

Also tried a variation of this and it looks stable, nice work
pics for reference

image
image
image
image
image
and the remote repo is still working perfectly 
image

Comment (cebtenzzre) [reply src/transformers/dynamic_module_utils.py:188]: > could you return the original parameter `repo_id` also the documentation for it too

The parameter was only added 10 hours ago in #29175, and there hasn't been a release since then. I don't think many people are already relying on this. And that change was merged without any deprecation period, so it doesn't seem like this is considered a stable public API.

**Review by Rocketknight1 — APPROVED** [review]
Strong approval!

For @ArthurZucker: This PR is fixing the same problem as my PR #29175. However, it has two advantages. Firstly, it's cleaner, rather than a messy try-except, and secondly it's more universal and doesn't fail in other cases when `.` is present in the repo path.

The only thing I'm confused about is that @cebtenzzre said that the test I added in #29175 was still passing even before that PR. That's strange to me - my test explicitly called `model = AutoModel.from_pretrained("hf-internal-testing/test_dynamic_model_v1.0", trust_remote_code=True)`, which should definitely have failed before my PR because the repo name contains `.` 

I'm not sure what's going on there - I'd appreciate any clarification from @cebtenzzre about what he thinks the cause is! Ideally we'd make sure that test is working correctly before we merge this.

**Comment by Rocketknight1** [comment]
Update: @not-lain explained the issue with the test. The repo being tested has now been updated, so that the test actually tests what we wanted to test in the first place, and that resolves my question!

@cebtenzzre I don't have any other objections, so we just need core maintainer review to merge this!

File: src/transformers/dynamic_module_utils.py
Comment (not-lain) [reply src/transformers/dynamic_module_utils.py:188]: @cebtenzzre i just rechecked, and yess it checks out, thanks a lot for clearing that out

**Review by ArthurZucker — APPROVED** [review]
A lot cleaner !

File: src/transformers/dynamic_module_utils.py
Comment (ali-tunahan) [inline src/transformers/dynamic_module_utils.py:194]: module_path now expects the file extension too now. Maybe that should be reflected here too? @cebtenzzre

**Pull request merged**
